runhouse/rns/run_module_utils.py

```diff
@@ -134,21 +190,300 @@ def get_fn_from_pointers(fn_pointers, serialize_res, num_gpus, *args, **kwargs):
+    )
+
+    config_path = resolve_absolute_path(run_config_file)
+
+    inputs = {"args": args, "kwargs": kwargs}
+
+    if not THIS_CLUSTER:
+        raise ValueError("Failed to get current cluster from config")
+
+    current_cluster: str = rh_config.rns_client.resolve_rns_path(THIS_CLUSTER)
+
+    # Create a new Run object, and save down its config data to the log folder on the cluster
+    new_run = run(
+        name=run_name,
+        fn=fn,
+        overwrite=True,
+        fn_name=fn.__name__,
+        system=current_cluster,
+    )
+
+    # Save down config for new Run
+    new_run._register_new_fn_run()
+
+    # Save down pickled inputs to the function for the Run
+    new_run.write(
+        data=pickle.dumps(inputs),
+        path=new_run._fn_inputs_path(),
+    )
+
+    logger.info(f"Finished writing inputs for {run_name} to path: {config_path}")
+
+    return new_run
+
+
+def _get_result_from_ray(obj_ref):
+    from runhouse.rns.run import RunStatus
+
+    try:
+        result = ray.get(obj_ref)
+        return result, RunStatus.COMPLETED
+
+    except Exception as e:
+        return str(e), RunStatus.ERROR
+
+
+def _register_completed_run(path, run_status):
+    from runhouse import Run
+
+    logger.info(f"Registering completed run in path: {path}")
+    run_obj = Run.from_path(path)
+    logger.info(f"Run obj in register completed run: {run_obj.config_for_rns}")
+
+    # Update the config data for the completed run
+    run_obj._register_fn_run_completion(run_status)
+
+    return run_obj
+
+
+def _existing_run_from_file(run_key):
+    from runhouse import Run
+
+    if run_key == "latest":
+        # TODO [JL]
+        raise NotImplementedError("Latest not currently supported")
+    else:
+        folder_path = Run._base_cluster_folder_path(run_key)
+
+    folder_path_on_system = resolve_absolute_path(folder_path)
+
+    existing_run = Run.from_path(path=folder_path_on_system)
+
+    return existing_run
+
+
+def fn_from_module_path(relative_path, fn_name, module_name):
+    module_path = (
+        str((Path.home() / relative_path).resolve()) if relative_path else None
+    )
+    logger.info(f"Module path on unary server: {module_path}")
+
+    if module_name == "notebook":
+        fn = fn_name  # Already unpickled above
+    else:
+        fn = get_fn_by_name(module_name, fn_name, module_path)
+
+    return fn
+
+
+def create_command_based_run(run_name, commands, cmd_prefix, python_cmd):
+    from runhouse import Run, run
+    from runhouse.rns.obj_store import THIS_CLUSTER
+    from runhouse.rns.run import RunStatus
+
+    folder_path = Run._base_cluster_folder_path(run_name)
+    folder_path_on_system = resolve_absolute_path(folder_path)
+
+    run_obj = run(
+        name=run_name, cmds=commands, overwrite=True, path=folder_path_on_system
+    )
+
+    run_obj._register_new_fn_run()
+
+    final_stdout = []
+    final_stderr = []
+
+    for command in commands:
```

Comment:
Downside of running via rpc here compared to with the CommandRunner is that we're not streaming logs back to the user.